zapisy/apps/grade/poll/views.py

```diff
@@ -183,88 +186,153 @@ def __get_counter_for_categories(polls):
         return number_of_submissions_for_category
 
     @staticmethod
-    def __get_processed_results(submissions):
+    def __are_read(polls, user):
+        is_read_category = defaultdict(True.__bool__)
+        is_read_poll = dict()
+
+        last_views: Dict[Poll, datetime.datetime] = dict(
+                PollView.objects.filter(user=user, poll__in=polls).values_list("poll", "time")
+            )
+
+        last_modifieds = dict(
+            Submission.objects.filter(poll__in=polls, submitted=True)
+            .order_by("poll", "modified")
+            .distinct("poll")
+            .values_list("poll", "modified")
+        )
+        for poll in polls:
+            is_read_poll[poll] = poll.id not in last_modifieds or (
+                poll.id in last_views and last_views[poll.id] > last_modifieds[poll.id]
+            )
+            is_read_category[poll.category] &= is_read_poll[poll]
+        return [is_read_category, is_read_poll]
+
+    @staticmethod
+    def __get_processed_results(current_poll, user, submissions):
         poll_results = PollSummarizedResults(
             display_answers_count=True, display_plots=True
         )
 
+        if not submissions:
+            return poll_results
+
+        try:
+            last_time = PollView.objects.get(user=user, poll=current_poll).time
+        except PollView.DoesNotExist:
+            last_time = None
+
         for submission in submissions:
-            if 'schema' in submission.answers:
-                for entry in submission.answers['schema']:
+            if 'schema' not in submission.answers:
+                continue
+            for entry in submission.answers['schema']:
+                if 'choices' in entry:
+                    choices = entry['choices']
+                    viewed = None
+                else:
                     choices = None
-                    if 'choices' in entry:
-                        choices = entry['choices']
-                    poll_results.add_entry(
-                        question=entry['question'],
-                        field_type=entry['type'],
-                        answer=entry['answer'],
-                        choices=choices,
-                    )
+                    if not last_time:
+                        viewed = False
+                    elif 'modified' in entry:
+                        viewed = dateutil.parser.isoparse(entry['modified']) < last_time
+                    else:
+                        viewed = submission.modified < last_time
```

Comment:
Ostatnia zmiana nieznacznie bo nieznacznie, ale jednak zmienia logikę. Konserwatywniej byłoby
```suggestion
                if 'choices' in entry:
                    choices = entry['choices']
                else:
                    choices = None
                
                if entry['type'] in ["radio", "checkbox"]:
                    viewed = None
                elif not last_time:
                    viewed = False
                elif 'modified' in entry:
                    viewed = dateutil.parser.isoparse(entry['modified']) < last_time
                else:
                    viewed = submission.modified < last_time
```